openmdao/core/group.py

```diff
@@ -246,28 +250,63 @@ def _setup_variables(self, compute_indices=False):
         self._sysdata._unknowns_dict = unknowns_dict
 
         self._data_xfer = {}
-        self._to_abs_unames = {}
-        self._to_abs_pnames = {}
+
+        to_prom = self._sysdata.to_prom = {}
+        to_abs_unames = self._sysdata.to_abs_unames = OrderedDict()
+        to_abs_pnames = self._sysdata.to_abs_pnames = OrderedDict()
+        to_prom_unames = self._sysdata.to_prom_unames = OrderedDict()
+        to_prom_pnames = self._sysdata.to_prom_pnames = OrderedDict()
 
         for sub in itervalues(self._subsystems):
             subparams, subunknowns = sub._setup_variables(compute_indices)
             for p, meta in iteritems(subparams):
-                meta = meta.copy()
-                meta['promoted_name'] = self._promoted_name(meta['promoted_name'], sub)
+                prom = self._promoted_name(sub._sysdata.to_prom_pnames[p], sub)
                 params_dict[p] = meta
-                self._to_abs_pnames.setdefault(meta['promoted_name'], []).append(p)
+                to_abs_pnames.setdefault(prom, []).append(p)
+                to_prom_pnames[p] = prom
 
             for u, meta in iteritems(subunknowns):
-                meta = meta.copy()
-                meta['promoted_name'] = self._promoted_name(meta['promoted_name'], sub)
+                prom = self._promoted_name(sub._sysdata.to_prom_unames[u], sub)
                 unknowns_dict[u] = meta
-                self._to_abs_unames.setdefault(meta['promoted_name'], []).append(u)
+                if prom in to_abs_unames:
+                    raise RuntimeError("'%s': promoted name '%s' matches "
+                                       "multiple unknowns: %s" %
+                                       (self.pathname, prom,
+                                        (to_abs_unames[prom], u)))
+
+                to_abs_unames[prom] = u
+                to_prom_unames[u] = prom
+
+            to_prom.update(to_prom_unames)
+            to_prom.update(to_prom_pnames)
 
             # check for any promotes that didn't match a variable
             sub._check_promotes()
 
         return self._params_dict, self._unknowns_dict
 
+    def _get_gs_outputs(self, mode, vois):
```

Comment:
a docstring (or even just a comment since its private) might be good here to explain this function